### PR TITLE
Make FeedbackSessionsResultsBundle resilient to errors in question numbers #4321

### LINKED ISSUE #4321: Make FeedbackSessionsResultsBundle resilient to errors in question numbers
When two questions have the same question number, they are treated as the same question by the result bundle. This is because TreeMap relies on <code>compareTo</code> to decide if a key already exist in the Map, and FeedbackQuestionAttributes compareTo uses only the questionNumber for comparison

As part of #4319, to allow the results page to load the responses even if the question numbers are wrong

### PR DESCRIPTION
**Pull request by kanghj**
Fixes #4321

Test behaviour: before the fix, the page loaded by the test will have a ClassCastException and fail to load.

### CODE REVIEW

File: src/test/resources/pages/instructorFeedbackResultsFilteredByNoSection.html [hunk 1]
```diff
@@ -690,7 +690,7 @@
                               <input name="frgroupbyteam" type="hidden" value="null"/>
                               <input name="frsorttype" type="hidden" value="question"/>
                               <input id="showStats-10" name="frshowstats" type="hidden" value="on"/>
-                              <input name="questionnum" type="hidden" value="10"/>
+                              <input name="questionnum" type="hidden" value="{*}"/>
                               <input name="frgroupbysection" type="hidden" value="None"/>
                            </form>
             <div class="display-icon pull-right">
```
Comment (damithc) [inline src/test/resources/pages/instructorFeedbackResultsFilteredByNoSection.html]: We can't predict the question number anymore?

File: src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html [hunk 2]
```diff
@@ -1,684 +1,701 @@
...
-                              <input name="frgroupbysection" type="hidden" value="All"/>
-                           </form>
-            <div class="display-icon pull-right">
-               <span class="glyphicon glyphicon-chevron-down pull-right">
-               </span>
+            <div class="panel panel-info">
+               <div class="panel-heading ajax_submit" data-target="#panelBodyCollapse-8" id="panelHeading-8" style="cursor: pointer;">
+                  <form action="/page/instructorFeedbackResultsPage" class="seeMoreForm-8" id="seeMore-8" style="display:none;">
+                     <input name="courseid" type="hidden" value="CFResultsUiT.CS2104"/>
+                                          <input name="fsname" type=
```
Comment (damithc) [inline src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html]: Looks like godmode changed more than it should. God mode is not fixed yet? @thyageshm

File: src/test/resources/pages/instructorFeedbackResultsFilteredByNoSection.html (hunk 1 above)
Comment (kanghj) [inline src/test/resources/pages/instructorFeedbackResultsFilteredByNoSection.html]: The parameter is the questionid now, will rename it.

**Comment by LowWeiLin** [comment]
Should there be code to detect and log when the question numbers are invalid?

**Comment by damithc** [comment]
I think so (do a severe log), but it can be a separate issue. In the current code, the user is not affected even if the question numbers are messed up?

File: src/test/resources/pages/instructorFeedbackResultsFilteredByNoSection.html (hunk 1 above)
Comment (thyageshm) [inline src/test/resources/pages/instructorFeedbackResultsFilteredByNoSection.html]: was this manually done or with godmode?

File: src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html (hunk 2 above)
Comment (thyageshm) [inline src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html]: Sir there seems to be a change from verifying just the "mainContent" div to the entire html page. This causes an indentation difference in everyline. Only if the html parts being compared are the same, the white space is ignored. One thing that can be done is to remove the indentation altogether. It reduces readability of the html pages, but will work in these kinds of scenarios as well (where we change the scope of comparison).

Comment (thyageshm) [inline src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html]: @kanghj, was there a change in the scope of comparison? I am not able to see a change in the test code associated with this page (it still says "verifyHtmlMainContent", but somehow the newly generated html has the entire html in it.)

Comment (kanghj) [inline src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html]: Actually, from what I can gather, there were a few html files that were missed out when you regenerated the html files.

**Comment by kanghj** [comment]
The effect of this branch is that it will allow the user to still see the question and responses, but the question number displayed will be wrong (e.g. all the questions will be numbered question 1)

File: src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html (hunk 2 above)
Comment (kanghj) [inline src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html]: Oh, I'm wrong, it looked like they were regenerated. I didn't change the scope of comparison.

File: src/test/resources/pages/instructorFeedbackResultsFilteredByNoSection.html (hunk 1 above)
Comment (kanghj) [inline src/test/resources/pages/instructorFeedbackResultsFilteredByNoSection.html]: This was done with godmode.

Comment (thyageshm) [inline src/test/resources/pages/instructorFeedbackResultsFilteredByNoSection.html]: @damithc sir, I believe godmode is working as expected. There seems to be a different kind of problem in the earlier cases and I am trying to resolve them with @kanghj

File: src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html (hunk 2 above)
Comment (thyageshm) [inline src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html]: Hmmm.... this is very weird. The same thing happens in the current master (it includes the html tags in the generated file). But when I try in AdminEmailPageUiTest, the generated file is ok

Comment (thyageshm) [inline src/test/resources/pages/instructorFeedbackResultsAjaxByQuestionViewForHelperOne.html]: I have found the error. Will fix it asap in #4324

**Comment by kanghj** [comment]
Updated. @LowWeiLin

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 3]
```diff
@@ -1076,9 +1076,11 @@ public String getQuestionText(String feedbackQuestionId) {
         }
 
         Map<FeedbackQuestionAttributes, List<FeedbackResponseAttributes>> sortedMap =
-                new TreeMap<FeedbackQuestionAttributes, List<FeedbackResponseAttributes>>();
+                new LinkedHashMap<FeedbackQuestionAttributes, List<FeedbackResponseAttributes>>();
 
-        for (FeedbackQuestionAttributes question : questions.values()) {
+        List<FeedbackQuestionAttributes> sortedQuestions = new ArrayList<>(questions.values());
+        Collections.sort(sortedQuestions);
```
Comment (LowWeiLin) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Can add a comment to say what it is sorted by?

Comment (LowWeiLin) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Or add it to the variable name.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 4]
```diff
@@ -66,7 +66,7 @@
     private static final InstructorsLogic instructorsLogic = InstructorsLogic.inst();
     private static final CoursesLogic coursesLogic = CoursesLogic.inst();
     private static final StudentsLogic studentsLogic = StudentsLogic.inst();
-    private static final int QUESTION_NUM_FOR_RESPONSE_RATE = -1;
+    private static final String QUESTION_NUM_FOR_RESPONSE_RATE = "-1";
```
Comment (LowWeiLin) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Variable name should be changed to question id now?

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 5]
```diff
@@ -1889,10 +1889,10 @@ private FeedbackSessionResultsBundle getFeedbackSessionResultsForUserWithParams(
                     visibilityTable, responseStatus, roster, responseComments);
         }
         
-        if (params.get("questionNum") != null) {
-            int questionNumber = Integer.parseInt(params.get("questionNum"));
+        if (params.get("questionId") != null) {
+            String questionId = params.get("questionId");
             FeedbackQuestionAttributes question = fqLogic.getFeedbackQuestion(
-                    feedbackSessionName, courseId, questionNumber);
+                                                                    questionId);
```
Comment (LowWeiLin) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: It does not make sense to get the feedback question if questionId is QUESTION_NUM_FOR_RESPONSE_RATE, then check for null later. Can create a separate issue to fix this logic.

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackResultsPageUiTest.java [hunk 6]
```diff
@@ -97,7 +98,31 @@ public void testContent() {
...
+        assertEquals(question, questionFromDatastore);
+        // make question have the same question number as another question
+        questionFromDatastore.questionNumber = 1;
+        
+        BackDoor.editFeedbackQuestion(questionFromDatastore);
 
+        resultsPage = loginToInstructorFeedbackResultsPage("CFResultsUiT.instr", "Session with errors");
+        
+        resultsPage.waitForPanelsToCollapse();
+        resultsPage.verifyHtmlMainContent("/instructorFeedbackResultsPageWithDuplicateQuestionNumber.html");
```
Comment (LowWeiLin) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackResultsPageUiTest.java]: Will this consistently pass? The sort order of the questions will not be deterministic now right?

Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackResultsPageUiTest.java]: Modified to verify html of each question panel separately

Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackResultsPageUiTest.java]: Oops, this is failing sometimes. The id of each question panel is still determined by sort order.

**Pull request merged**